3-enrich/spark-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich.spark/EnrichJob.scala

```diff
@@ -0,0 +1,173 @@
+  @transient private val sc: SparkContext = spark.sparkContext
+  import spark.implicits._
+
+  // Necessary to be able to deal with LZO-compressed files
+  @transient private val hadoopConfig = sc.hadoopConfiguration
+  hadoopConfig.set("io.compression.codecs", classOf[com.hadoop.compression.lzo.LzopCodec].getName())
+  hadoopConfig.set("io.compression.codec.lzo.class",
+    classOf[com.hadoop.compression.lzo.LzoCodec].getName())
+
+  // Job configuration
+  private val enrichConfig = EnrichJobConfig.loadConfigFrom(args).fold(
+    e => throw FatalEtlError(e.map(_.toString)),
+    identity
+  )
+
+  def run(): Unit = {
+    import EnrichJob._
+    import singleton._
+
+    // Install MaxMind file(s) if we have them
+    enrichConfig.filesToCache.foreach { case (uri, filename) => sc.addFile(uri.toString) }
+
+    val inFormat = enrichConfig.inFormat
+    val igluConfig = enrichConfig.igluConfig
+
+    val input = getInputRDD(inFormat, enrichConfig.inFolder)
+
+    val common = input
+      .map { line =>
+        (
+          line,
+          EtlPipeline.processEvents(
+            RegistrySingleton.get(igluConfig, enrichConfig.enrichments, enrichConfig.local),
+            etlVersion,
+            enrichConfig.etlTstamp,
+            LoaderSingleton.get(inFormat).asInstanceOf[Loader[Any]].toCollectorPayload(line)
+          )(ResolverSingleton.get(igluConfig))
+        )
+      }
+      .cache()
+
+    // Handling of malformed rows
+    val bad = common
+      .map { case (line, enriched) => (line, projectBads(enriched)) }
+      .flatMap { case (line, errors) =>
+        val originalLine = line match {
+          case bytes: Array[Byte] => new String(Base64.encodeBase64(bytes), "UTF-8")
+          case other => other.toString
+        }
+        errors.map(BadRow(originalLine, _).toCompactJson)
+      }
+    if (!bad.isEmpty) bad.saveAsTextFile(enrichConfig.badFolder)
+
+    // Handling of properly-formed rows
+    val good = common
+      .flatMap { case (line, enriched) => projectGoods(enriched) }
+    if (!good.isEmpty)
+      spark.createDataset(good)(Encoders.bean(classOf[EnrichedEvent]))
+        .toDF()
+        // hack to preserve the order of the fields in the csv, otherwise it's alphabetical
+        .select(classOf[EnrichedEvent].getDeclaredFields().map(f => col(f.getName())): _*)
+        .write
+        .option("sep", "\t")
+        .option("escape", "")
+        .option("quote", "")
+        .csv(enrichConfig.outFolder)
+  }
+
+  /**
+   * Read the input files.
+   * @param inFormat Collector format in which the data is coming in
+   * @return A RDD containing strings or byte arrays
+   */
+  private def getInputRDD(inFormat: String, path: String): RDD[_] = {
+    inFormat match {
+      case "thrift" =>
+        import com.twitter.elephantbird.mapreduce.input.MultiInputFormat
+        MultiInputFormat.setClassConf(classOf[Array[Byte]], hadoopConfig)
+        sc.newAPIHadoopFile[
+          org.apache.hadoop.io.LongWritable,
```

Comment:
Well, maybe this is matter of taste, but I'd still import things that are used only once at the top of file.